src/librbd/Operations.cc

```diff
@@ -1548,20 +1547,17 @@ int Operations<I>::metadata_remove(const std::string &key) {
   if(r < 0)
     return r;
 
-  C_SaferCond metadata_ctx;
-  {
-    std::shared_lock owner_lock{m_image_ctx.owner_lock};
-    r = prepare_image_update(exclusive_lock::OPERATION_REQUEST_TYPE_GENERAL,
-                             true);
-    if (r < 0) {
-      return r;
-    }
-
-    execute_metadata_remove(key, &metadata_ctx);
+  r = invoke_async_request("metadata_remove",
+                           exclusive_lock::OPERATION_REQUEST_TYPE_GENERAL,
+                           true,
+                           boost::bind(&Operations<I>::execute_metadata_remove,
+                                       this, key, _1),
+                           boost::bind(&ImageWatcher<I>::notify_metadata_remove,
+                                       m_image_ctx.image_watcher, key, _1));
+  if (r < 0 && r != -EINVAL) {
```

Comment:
Nit: can't remember, but will this return `-ENOENT` if the key doesn't exist? If so, since RPC messages can be replayed you should filter out `-ENOENT` here (instead of `-EINVAL`) and add a quick check that the key exists before sending the RPC.

Reply:
No, I think, if the key doesn't exist, `ENOENT` is returned at [line1548](https://github.com/ceph/ceph/pull/37121/files#diff-5b3b92b8fcd4cfc8b36ddf37be5006e3R1548).